### PR TITLE
csi: make leader election flags configurable

### LINKED ISSUE #13475: RBD provisioner stuck attempting to acquire leader lease for 30s
It's taking over 30s for a leader lease to be acquired and for an RBD to be provisioned, this is on rook 1.13 on a healthy cluster rke2 kubernetes v1.29

### PR DESCRIPTION
**Pull request by Madhu-1**
This commit adds the flexibility to configure leader election flags per the user requirement and keeps the existing values as the default ones.

**Checklist:**

- [ ] **Commit Message Formatting**: Commit titles and messages follow guidelines in the [developer guide](https://rook.io/docs/rook/latest/Contributing/development-flow/#commit-structure).
- [ ] Reviewed the developer guide on [Submitting a Pull Request](https://rook.io/docs/rook/latest/Contributing/development-flow/#submitting-a-pull-request)
- [ ] [Pending release notes](https://github.com/rook/rook/blob/master/PendingReleaseNotes.md) updated with breaking and/or notable changes for the next minor release.
- [ ] Documentation has been updated, if necessary.
- [ ] Unit tests have been added, if necessary.
- [ ] Integration tests have been added, if necessary.

### CODE REVIEW

File: Documentation/Helm-Charts/operator-chart.md [hunk 1]
```diff
@@ -70,6 +70,9 @@ The following table lists the configurable parameters of the rook-operator chart
 | `csi.csiCephFSPluginVolume` | The volume of the CephCSI CephFS plugin DaemonSet | `nil` |
 | `csi.csiCephFSPluginVolumeMount` | The volume mounts of the CephCSI CephFS plugin DaemonSet | `nil` |
 | `csi.csiCephFSProvisionerResource` | CEPH CSI CephFS provisioner resource requirement list | see values.yaml |
+| `csi.csiLeaderElectionLeaseDuration` | Duration in seconds that non-leader candidates will wait to force acquire leadership. | `137s` |
+| `csi.csiLeaderElectionRenewDeadline` | Deadline in seconds that the acting leader will retry refreshing leadership before giving up. | `107s` |
+| `csi.csiLeaderElectionRetryPeriod` | Retry period in seconds the LeaderElector clients should wait between tries of actions. | `26s` |
```
Comment (subhamkrai) [inline Documentation/Helm-Charts/operator-chart.md:75]:
```suggestion
| `csi.csiLeaderElectionLeaseDuration` | Duration in seconds that non-leader candidates will wait to force acquire leadership. | `137s` |
| `csi.csiLeaderElectionRenewDeadline` | Deadline in seconds that the acting leader will retry refreshing leadership before giving up. | `107s` |
| `csi.csiLeaderElectionRetryPeriod` | Retry period in seconds the LeaderElector clients should wait between tries of actions. | `26s` |
```

how these values are taken? On this doc I see values are quite lesshttps://github.com/kubernetes-csi/external-provisioner?tab=readme-ov-file#command-line-options

Comment (Madhu-1) [reply Documentation/Helm-Charts/operator-chart.md:75]: @subhamkrai These are the values suggested by the openshift storage team long back, btw what the suggested change I cannot see any.

Comment (subhamkrai) [reply Documentation/Helm-Charts/operator-chart.md:75]:
```text
--leader-election-lease-duration <duration>: Duration, in seconds, that non-leader candidates will wait to force acquire leadership. Defaults to 15 seconds.

--leader-election-renew-deadline <duration>: Duration, in seconds, that the acting leader will retry refreshing leadership before giving up. Defaults to 10 seconds.

--leader-election-retry-period <duration>: Duration, in seconds, the LeaderElector clients should wait between tries of actions. Defaults to 5 seconds.```

Comment (subhamkrai) [reply Documentation/Helm-Charts/operator-chart.md:75]: The defaults value are 10s,10s,5s repectively

Comment (Madhu-1) [reply Documentation/Helm-Charts/operator-chart.md:75]: @subhamkrai i dont have any plan to change the default values what we have in Rook from very long time and its working fine without any issue. https://github.com/rook/rook/blob/b9fa0f44d5dd00bdbcc99097338674a1fd55815f/pkg/operator/ceph/csi/template/rbd/csi-rbdplugin-provisioner-dep.yaml#L34-L36 these are the same values we have release 1.10. This PR is exposing user options to change the values if they want and doesn't change any existing values

Comment (subhamkrai) [reply Documentation/Helm-Charts/operator-chart.md:75]: Okay, got it.

**Review by subhamkrai — APPROVED** [review]

File: pkg/operator/ceph/csi/spec.go [hunk 2]
```diff
@@ -235,6 +238,11 @@ const (
 	defaultLogLevel        uint8 = 0
 	defaultSidecarLogLevel uint8 = 0
 
+	// default leader election flags
+	defaultLeaderElectionLeaseDuration = 137 * time.Second
```
Comment (parth-gr) [inline pkg/operator/ceph/csi/spec.go:242]: how these times are justified?

Comment (subhamkrai) [reply pkg/operator/ceph/csi/spec.go:242]: @parth-gr I asked the same question in the resolved comment https://github.com/rook/rook/pull/13573#discussion_r1453319924 :Xd

Comment (parth-gr) [reply pkg/operator/ceph/csi/spec.go:242]: I see they were the defaults before

File: pkg/operator/ceph/csi/csi.go [hunk 3]
```diff
@@ -238,6 +238,39 @@ func (r *ReconcileCSI) setParams(ver *version.Info) error {
 		}
 	}
 
+	leaderElectionLeaseDuration := k8sutil.GetValue(r.opConfig.Parameters, "CSI_LEADER_ELECTION_LEASE_DURATION", "")
```
Comment (parth-gr) [inline pkg/operator/ceph/csi/csi.go:241]: should we add a upper and lower limit to them?

Comment (parth-gr) [reply pkg/operator/ceph/csi/csi.go:241]: correct, I just wanted to make sure that the user won't add anything that makes the cluster down for a long time and not something less so the cluster cannot select the leader,
Btw would like to pass any doc from CSI that explains how this leader is decided.

Comment (Madhu-1) [reply pkg/operator/ceph/csi/csi.go:241]: This is advance setting and its time.Duration flag, if invalid values are passed csi provisioner pod will be down it wont affect the whole cluster or any existing workloads and i assume its admin's responsibility to choose the correct value. https://kubernetes-csi.github.io/docs/introduction.html contains all the docs.

**Comment by travisn** [comment]
Why would these values need to be customizable? We have been using these defaults for so long, it's surprising now to make them configurable.

**Comment by travisn** [comment]
> Why would these values need to be customizable? We have been using these defaults for so long, it's surprising now to make them configurable.

I see now the discussion in #13475

**Pull request merged**